kernel/vmemitcodes.ml

```diff
@@ -54,13 +54,74 @@ module RelocTable = Hashtbl.Make(struct
   let hash = hash_reloc_info
 end)
 
+module Positions :
+sig
+  type t
+  val of_list : int list -> t
+  val iter : (int -> unit) -> t -> unit
+end =
+struct
+
+type t = string
+(* Represent an ordered set of 32-bit integers as an array of successive diffs.
+   We use furthermore an approach where smaller integers use less bytes. Numbers
+   smaller than 255 are stored into one byte. This is a cheap way to compact
+   this data. *)
+
+let output buf n =
+  if n < 0xFF then Buffer.add_uint8 buf n
+  else
+    let () = Buffer.add_uint8 buf 0xFF in
+    Buffer.add_int32_be buf (Int32.of_int n)
+
+let input s pos =
+  let c = Char.code s.[!pos] in
+  if c < 0xFF then
+    let () = pos := !pos + 1 in
+    c
+  else
+    let c1 = Char.code s.[!pos + 1] in
+    let c2 = Char.code s.[!pos + 2] in
+    let c3 = Char.code s.[!pos + 3] in
+    let c4 = Char.code s.[!pos + 4] in
+    let () = pos := !pos + 5 in
+    (c1 lsl 24) lor (c2 lsl 16) lor (c3 lsl 8) lor c4
+
+let of_list l = match l with
+| [] -> ""
+| n :: l ->
+  let buf = Buffer.create 16 in
+  let () = output buf n in
+  let rec aux cur l = match l with
+  | [] -> ()
+  | n :: l ->
+    let () = assert (cur < n) in
+    let () = output buf (n - cur) in
+    aux n l
+  in
+  let () = aux n l in
+  Buffer.contents buf
+
+let iter f s =
+  let pos = ref 0 in
+  let len = String.length s in
+  let cur = ref 0 in
+  while !pos < len do
+    let n = input s pos in
+    let () = cur := n + !cur in
+    f !cur
+  done
+
+end
+
 (** We use arrays for on-disk representation. On 32-bit machines, this means we
     can only have a maximum amount of about 4.10^6 relocations, which seems
     quite a lot, but potentially reachable if e.g. compiling big proofs. This
     would prevent VM computing with these terms on 32-bit architectures. Maybe
     we should use a more robust data structure? *)
 type patches = {
-  reloc_infos : (reloc_info * int array) array;
+  reloc_infos : reloc_info array;
```

Comment:
The comment above needs to be updated, since the number of relocations is no longer bounded.

(The number of _distinct_ relocation targets is still bounded, for the same reason as before. But I do not know if Coq actually supports handling 4 millions of those on a 32-bit architecture anyway, since they need to exist in the first place, irrespective of the VM.)

Reply:
Replying to myself, this is indeed impossible. And in fact, the whole comment was already meaningless for the old code, since you cannot have more relocations in a bytecode string than the number of 32-bit words in a bytecode string, by definition.